### PR TITLE
Enhanced usage of MPI_abort

### PR DESCRIPTION
**Pull request by fverdugo**
This should fix https://github.com/gridap/GridapPETSc.jl/issues/71

cc @amartinhuertas

### CODE REVIEW

**Comment by fverdugo** [comment]
@amartinhuertas can you please review this?

**Review by amartinhuertas — CHANGES_REQUESTED** [review]
All good, except for the following I am not sure. In particular, I think we should update the following line 

https://github.com/fverdugo/PartitionedArrays.jl/blob/master/test/mpi/mpiexec.jl#L13

and test for the code error returned by the program, right? (i..e, is it 0 or not?)

**Comment by fverdugo** [comment]
> All good, except for the following I am not sure. In particular, I think we should update the following line
> https://github.com/fverdugo/PartitionedArrays.jl/blob/master/test/mpi/mpiexec.jl#L13
> and test for the code error returned by the program, right? (i..e, is it 0 or not?)

The fundamental change in this PR is to call MPI_Abort with a non-zero error code. 

Calling it with a 0 error code (as before) was causing the problems related with https://github.com/gridap/GridapPETSc.jl/issues/71

The `run` command  was thinking that  `mpiexec` was returning smoothly, even when calling MPI_Abort (since we returned error code 0).

In contrast, if we call MPI_Abort with a non-zero code, `run` detects that the launched process failed and raises an exception. As a result the computation is stopped, line  https://github.com/fverdugo/PartitionedArrays.jl/blob/master/test/mpi/mpiexec.jl#L13 is not reached, and the CI environment receives a non-zero error code (caused by the exception lauched by `run`) which signals that the tests have failed (as we want).

I think that the PR is fine, just wanted to double check. 

It makes sense now for you?

**Comment by amartinhuertas** [comment]
> The fundamental change in this PR is to call MPI_Abort with a non-zero error code.

Yes, that I understood. 

> In contrast, if we call MPI_Abort with a non-zero code, run detects that the launched process failed and raises an exception. As a result the computation is stopped, line https://github.com/fverdugo/PartitionedArrays.jl/blob/master/test/mpi/mpiexec.jl#L13 is not reached, and the CI environment receives a non-zero error code (caused by the exception lauched by run) which signals that the tests have failed (as we want).

Ok. This is what I was missing from the discussion.  Still I find confusing test true. I would do test error_code==0, even if we know that error code is going to be always equal to zero (otherwise the execution is aborted).

**Comment by fverdugo** [comment]
>  I would do test error_code==0, even if we know that error code is going to be always equal to zero (otherwise the execution is aborted).

I see your point, but unfortunately, finding the error code would complicate the code.

I would keep it like this for simplicity. In fact, I don't find it so confusing. It is quite aparent that `@test true` will be reached only if `run` succeeds, right?

**Comment by amartinhuertas** [comment]
> I would keep it like this for simplicity. In fact, I don't find it so confusing. It is quite aparent that @test true will be reached only if run succeeds, right?

Ok. Put at comment then. I dont think is quite aparent.

**Comment by fverdugo** [comment]
> Ok. Put at comment then.
Done

**Review by amartinhuertas — APPROVED** [review]

File: test/mpi/mpiexec.jl [hunk 1]
```diff
@@ -10,6 +10,8 @@ function run_mpi_driver(;procs,file)
     else
       run(`$cmd -n $procs $(Base.julia_cmd()) --project=$repodir $(joinpath(mpidir,file))`)
     end
-    @test true
+    # This line will be reached if and only if the command launched by `run` runs without errors.
+    # Then, if we arrive here, the test has succeeded.
+    @test true
```
Comment (fredrikekre) [inline test/mpi/mpiexec.jl:15]: Just FYI, there is the `success` function (https://docs.julialang.org/en/v1/base/base/#Base.success) for just making sure that a command returns 0, so the `run` calls above could be replaced with e.g.
```
@test success(`$cmd -n ...`)
```

Comment (fverdugo) [reply test/mpi/mpiexec.jl:15]: Very good point!

**Pull request merged**